src/graph.rs

```diff
@@ -255,6 +255,17 @@ pub enum Value {
 }
 
 impl Value {
+    /// Coerces this value into a boolean, returning an error if it's some other type of value.
+    pub fn into_bool(self, graph: &Graph) -> Result<bool, ExecutionError> {
+        match self {
+            Value::Boolean(value) => Ok(value),
```

Comment:
I'm sorely tempted to suggest allowing coercing `#null` into `False` and any `tree-sitter` syntax node into `True`. This would make it a bit more ergonomic when checking for the presence of captures, and avoid a whole bunch of `(is-null ...)` calls.

Reply:
I understand why it would be nice. I think a decision on that depends on what we expect/hope to do in the future in terms of static checking of DSL files. Such convenience features, which are easy to support in an untyped language, can really hurt when you later want to add checking. That is why I am usually quite conservative about them.

On the other hand, since optional captures are an integral part of the queries, this warrants special syntax, I suppose. Would a `?` postfix operator satisfy your temptation?

Reply:
I agree about this potentially harming the kind of static analysis we can do. I tried my best to think of a case where this (i.e. a capture silently being coerced into a boolean) might introduce a subtle bug, but I couldn't come up with a convincing example of this.

I think the postfix `?` syntax is likely sufficiently ergonomic.

Speaking of type safety (and being conservative), we should consider using a special `Value::NoCapture` constant instead of `Value::Null`. Then we could limit the `?` operator to just captures until such a time as we find a more general need for it. (I'm sure a nullness check _could_ have other uses, but I can't think of any at the moment.)

Reply:
I agree about this potentially harming the kind of static analysis we can do. I tried my best to think of a case where this (i.e. a capture silently being coerced into a boolean) might introduce a subtle bug, but I couldn't come up with a convincing example of this.

I think the postfix `?` syntax is likely sufficiently ergonomic.

Speaking of type safety (and being conservative), we should consider using a special `Value::NoCapture` constant instead of `Value::Null`. Then we could limit the `?` operator to just captures until such a time as we find a more general need for it. (I'm sure a nullness check _could_ have other uses, but I can't think of any at the moment.)

Reply:
> I tried my best to think of a case where this (i.e. a capture silently being coerced into a boolean) might introduce a subtle bug

I also could not come up with one, so it's more an argument of potential type checker complexity than correctness problems.

I'll leave it as this for the moment, and poke @dcreager to discuss some guidelines for how to approach these decisions in the future.

Reply:
After thinking about your comments some more, I have developed more concerns for a future typing approach. Assume optional values to have a `Option<T>` type. We now use `_? : Option<T> -> bool` to test if it is zero. But we use the `Option<T>` in a `T` position without an explicit operation to without explicitly converting `Option<T> -> T`. So there is still implicit coercion (or some form of flow typing) involved here.

Assuming we want option to be properly type checked, a "simple" type checker would require explicit operations for `is-null` as well as `unwrap`. Any setup where any one of these is implicit would require flow typing, coercion, (maybe other options?), or just not representing nullability in the types.

Reply:
I guess ideally we would have that the following code
```
if @capture? { ... }
```
behaves as the following pseudo-Rust-code.
```rust
if let Some(@capture) = @capture { ... }
```
In other words, we match _and rebind_ `@capture` based on whether it's `Some` or `None`. That way, we don't have to explicitly `unwrap` it inside the `if`.

That would make the coercion explicit, though it would also mean that there's no way to do complex boolean tests with captures. :disappointed: 

I see your point about type checking, though. I'm assuming "flow typing" refers to the fact that it's safe to access `@capture` if we're inside a conditional where the condition implies `@capture` cannot be empty. The trouble here, of course, is that it may be difficult to determine whether this implication is actually true.

On its own, I don't see how `unwrap` helps, since that surely just punts the checking to runtime instead. Likewise for the implicit coercion.

Of course, there is also the option (which may be the same as your final option) that is more in line with what I suggested in #15: make the `?` operator something special that inspects the local environment for the existence of a given capture and returns a boolean based on whether it's there or not. Then `(_)? @capture` can behave as it does currently, either setting `@capture` to the captured syntax node, or _not_ setting it if the match is empty. No option types would be needed then.

I guess that's also undesirable, though. Ultimately, we would want to know that all captures either _must_ exist, or if they don't, that this has been accounted for through control-flow.

This all makes me think that putting the control flow _inside_ the stanzas is perhaps a bad idea (apart from `scan` which is just weird, but in a very local way). Instead, maybe we should push this checking of capture existence to the top level of the stanza -- or rather, allow multiple, guarded sets of actions for a single query. 

So, something like
```
(foo
  bar: (_)? @bar
  baz: (_)? @baz
  quux: (_) @quux
) @foo
{ ; Default stanza, always run
  ... ; we can freely refer to @foo and @quux in here
}
(@bar & @baz) { ; only run if @bar and @baz are both defined
  ... ; we can freely refer to @foo, @bar, @baz, and @quux in here
}
(@bar & !baz) { ; only run if @bar is defined and @baz is not
  ... ; we can freely refer to @foo, @bar, and @quux in here
}
...
```
Crucially, we would only allow boolean expressions over captures, so figuring out whether something is implied should be straightforward (and in practice, I don't expect these expressions to be terribly complicated). (In fact, unless I'm mistaken, these "guards" would essentially just reduce to conjunctions of possibly negated captures.)

I apologise for turning this into a sort of meandering comment. To be honest, I would be fine with just picking an option (like this PR) for now, and then later on fixing things up to be more robust.

Reply:
> I apologise for turning this into a sort of meandering comment.

Not at all, you raise some interesting points!

> To be honest, I would be fine with just picking an option (like this PR) for now, and then later on fixing things up to be more robust.

I think that is a good idea.